File: node_worker/stats.js
export function calculateStatistics(samples) {
  if (!samples || samples.length === 0) {
    return {
      min: null,
      max: null,
      mean: null,
      median: null,
      q1: null,
      q3: null,
      standard_deviation: null,
    };
  }

  const sorted = [...samples].map(Number).sort((a, b) => a - b);
  const n = sorted.length;
  const min = sorted[0];
  const max = sorted[n - 1];
  const mean = sorted.reduce((sum, v) => sum + v, 0) / n;
  const median = n % 2 === 1 ? sorted[Math.floor(n / 2)] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2;

  const idx = (fraction) => {
    const i = Math.max(Math.min(Math.ceil(fraction) - 1, n - 1), 0);
    return i;
  };

  const q1 = sorted[idx(n / 4)];
  const q3 = sorted[idx((3 * n) / 4)];

  const variance = sorted.reduce((acc, value) => {
    const diff = value - mean;
    return acc + diff * diff;
  }, 0) / n;
  const standardDeviation = Math.sqrt(variance);

  return {
    min,
    max,
    mean,
    median,
    q1,
    q3,
    standard_deviation: standardDeviation,
  };
}

File: golang_worker/stats.go
package main

import (
    "math"
    "sort"
)

type Stats struct {
    Min    float64
    Max    float64
    Mean   float64
    Median float64
    Q1     float64
    Q3     float64
    StdDev float64
}

func calculateStatistics(samples []float64) Stats {
    if len(samples) == 0 {
        return Stats{}
    }
    s := append([]float64(nil), samples...)
    sort.Float64s(s)
    n := len(s)
    min := s[0]
    max := s[n-1]

    mean := 0.0
    for _, v := range s {
        mean += v
    }
    mean /= float64(n)

    var median float64
    if n%2 == 1 {
        median = s[n/2]
    } else {
        median = (s[n/2-1] + s[n/2]) / 2.0
    }

    idx := func(f float64) int {
        i := int(math.Ceil(f) - 1)
        if i < 0 {
            i = 0
        }
        if i >= n {
            i = n - 1
        }
        return i
    }
    q1 := s[idx(float64(n)/4.0)]
    q3 := s[idx(float64(3*n)/4.0)]

    var sumsq float64
    for _, v := range s {
        d := v - mean
        sumsq += d * d
    }
    variance := sumsq / float64(n)
    stddev := math.Sqrt(variance)

    return Stats{Min: min, Max: max, Mean: mean, Median: median, Q1: q1, Q3: q3, StdDev: stddev}
}

